Detectors/Calibration/workflow/CCDBPopulatorSpec.h

```diff
@@ -0,0 +1,96 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef O2_CALIBRATION_CCDBPOPULATOR_H
+#define O2_CALIBRATION_CCDBPOPULATOR_H
+
+/// @file   CCDBPopulator.h
+/// @brief  device to populate CCDB
+
+#include "Framework/DeviceSpec.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/ControlService.h"
+#include "Framework/WorkflowSpec.h"
+#include "Framework/Task.h"
+#include "Framework/CompletionPolicyHelpers.h"
+#include "CCDB/BasicCCDBManager.h"
+#include "CCDB/CcdbApi.h"
+#include "DetectorsCalibration/CalibObjectWrapper.h"
+#include "DataFormatsTOF/CalibLHCphaseTOF.h"
+
+using CcdbManager = o2::ccdb::BasicCCDBManager;
+using LhcPhase = o2::dataformats::CalibLHCphaseTOF;
+
+namespace o2
+{
+namespace calibration
+{
+  using LHCphase = o2::dataformats::CalibLHCphaseTOF;
+  using CalibObjectWrapper = o2::calibration::CalibObjectWrapper;
+  using LHCphaseVector = std::vector<LHCphase>;
+  using CalibObjectWrapperVector = std::vector<CalibObjectWrapper>;
+  using CcdbApi = o2::ccdb::CcdbApi;
+  
+class CCDBPopulator : public o2::framework::Task
+{
+ public:
+  void init(o2::framework::InitContext& ic) final
+  {    
+    mCCDBpath = ic.options().get<std::string>("ccdb-path");
+    auto& mgr = CcdbManager::instance();
+    mgr.setURL(mCCDBpath);
+  }
+
+  void run(o2::framework::ProcessingContext& pc) final
+  {
+    auto& mgr = CcdbManager::instance();
+    CcdbApi api;
+    api.init(mgr.getURL());
+
+    const auto& lhcclockoutput = pc.inputs().get<LHCphaseVector>("lhcphasecalib");
```

Comment:
should not we have here
````
 if (lhcclockwrapoutput[0].getObjectType() == ObjectType::lhcphase) {
   const auto& lhcclockoutput = pc.inputs().get<LHCphaseVector>("lhcphasecalib");
    for (int i = 0 ; i < lhcclockoutput.size(); i++) {
       auto lhcphasewrap = lhcclockwrapoutput[i];
        auto lhcphase = lhcclockoutput[i];
       api.storeAsTFileAny(&lhcphase, lhcphasewrap.getPath(), lhcphasewrap.getMetaData(), lhcphasewrap.getStartValidityTimestamp(), lhcphasewrap.getEndValidityTimestamp());
    }
  }
````

Reply:
Actually I am not sure here, I was thinking about it, but then I thought that if I find the object "lhcclockwrapoutput" (or "lhcphasecalib"), then it means that the LHCphase was sent (with its wrapper), no? So the check on the ObjectType could be not needed (and as such, the ObjectType itself).

Reply:
Right now the populator gets only LHCPhase, so you know what is received, but in general, how do you know a priori what was sent? We 1st need to check the ObjectType of the wrapper (actually, of the 1st element of this vector, since all others will be the same) to know what type we use in the  ``pc.inputs().get<...``

Reply:
Yes but also the wrapper is identified as being the one from the lhcphase. So I already know it is that one. Unless all wrappers are called generically. Maybe i am overlooking something...

Reply:
Yes but also the wrapper is identified as being the one from the lhcphase. So I already know it is that one. Unless all wrappers are called generically. Maybe i am overlooking something...

Reply:
No, the wrapper is a generic class (I would actually suppress any reference on specific object type, see other comment below), it just has an identifier ObjectType which allows you to know that the type of the shipped object. 
Each time this DPL device receives some input, the ``run`` method is triggerered. If the
received object is not an LHCPhase but you treet is as such, you will get an error.

Reply:
I also thought of a generic class, but I was not sure how to do it in the workflow. I was imagining several devices sending a wrapper, I think one needs a way to distinguish them. As written later in this thread, let's see with Giulio how to deal with this, then I will change accordingly. Also teh fitting might not be the best one, for now, it is to try the flow - which I still have to do as you know :-)